src/ButteredBread.sol

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.25;
+
+import {OwnableUpgradeable} from "openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import {ERC20VotesUpgradeable} from
+    "openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC20VotesUpgradeable.sol";
+import {IERC20} from "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import {IButteredBread} from "src/interfaces/IButteredBread.sol";
+
+/**
+ * @title Breadchain Buttered Bread
+ * @notice Deposit Butter (LP tokens) to earn scaling rewards
+ * @author Breadchain Collective
+ * @custom:coauthor @RonTuretzky
+ * @custom:coauthor @daopunk
+ */
+contract ButteredBread is ERC20VotesUpgradeable, OwnableUpgradeable, IButteredBread {
+    /// @notice Access control for Breadchain sanctioned liquidity pools
+    mapping(address lp => bool allow) public allowlistedLPs;
+    /// @notice How much ButteredBread should be minted for a Liquidity Pool token (Butter)
+    mapping(address lp => uint256 factor) public scalingFactors;
+    /// @notice Butter balance by account and Liquidity Pool token deposited
+    mapping(address account => mapping(address lp => uint256 balance)) public accountToLPBalances;
+
+    modifier onlyAllowed(address _lp) {
+        if (allowlistedLPs[_lp] != true) revert NotAllowListed();
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @param _initData see IButteredBread
+    function initialize(InitData calldata _initData) external initializer {
+        if (_initData.liquidityPools.length != _initData.scalingFactors.length) revert InvalidValue();
+        __Ownable_init(msg.sender);
+        __ERC20_init(_initData.name, _initData.symbol);
+        for (uint256 i; i < _initData.liquidityPools.length; ++i) {
+            allowlistedLPs[_initData.liquidityPools[i]] = true;
+            scalingFactors[_initData.liquidityPools[i]] = _initData.scalingFactors[i];
+        }
+    }
+
+    /// @notice Deposit LP tokens
+    function deposit(address _lp, uint256 _amount) external virtual onlyAllowed(_lp) {
+        _deposit(msg.sender, _lp, _amount);
+    }
+
+    /// @notice Withdraw LP tokens
+    function withdraw(address _lp, uint256 _amount) external virtual {
+        _withdraw(msg.sender, _lp, _amount);
+    }
+
+    /// @notice allow or deny LP token
+    function modifyAllowList(address _lp, bool _allowed) external virtual onlyOwner {
+        allowlistedLPs[_lp] = _allowed;
+    }
+
+    /// @notice set LP token scaling factor
+    function modifyScalingFactor(address _lp, uint256 _factor) external virtual onlyOwner onlyAllowed(_lp) {
```

Comment:
we would need to convert the existing values in `accountToLPBalances` if this value was ever updated, otherwise users will get funds locked because `_burn(_account, _amount * scalingFactors[_lp]);` in `_withdraw` will try to burn more than they have if the scaling factor is decrease. It will also allow users to withdraw more than they put in, if the scaling factor is reduced.

Reply:
So if we change the scaling factor, we should basically do a full redeem of those lp tokens?

Reply:
added `_syncVotingWeight` function:

- called upon withdrawal to adjust burn amount
- can be called from governor during cast vote to update user balance
- can be batch called by anyone to update an array of users at one time